ui/src/app/machines/views/MachineDetails/MachineStorage/FilesystemsTable/FilesystemsTable.tsx

```diff
@@ -1,203 +1,282 @@
 import { useState } from "react";
 
 import { Button, MainTable, Tooltip } from "@canonical/react-components";
-import { useDispatch } from "react-redux";
+import { useDispatch, useSelector } from "react-redux";
 
 import ActionConfirm from "../ActionConfirm";
-import type { NormalisedFilesystem } from "../types";
-import { formatSize } from "../utils";
+import { formatSize, isMounted } from "../utils-new";
 
 import AddSpecialFilesystem from "./AddSpecialFilesystem";
 
 import TableMenu from "app/base/components/TableMenu";
+import type { TSFixMe } from "app/base/types";
 import { actions as machineActions } from "app/store/machine";
-import type { Machine, MachineStatus } from "app/store/machine/types";
+import machineSelectors from "app/store/machine/selectors";
+import type {
+  Disk,
+  Filesystem,
+  Machine,
+  Partition,
+} from "app/store/machine/types";
+import type { RootState } from "app/store/root/types";
 
 type Expanded = {
-  content: "removeFilesystem";
-  id: number;
+  content: "remove";
+  id: string;
 };
 
 type Props = {
   canEditStorage: boolean;
-  filesystems: NormalisedFilesystem[];
   systemId: Machine["system_id"];
 };
 
-/**
- * Generate the actions that a given filesystem can perform.
- * @param filesystem - the filesystem to check.
- * @param setExpanded - function to set the expanded table row and content.
- * @returns list of action links.
- */
-const getActionLinks = (
-  filesystem: NormalisedFilesystem,
-  setExpanded: (expanded: Expanded) => void
+const normaliseRowData = (
+  rowId: string,
+  fs: Filesystem,
+  storageDevice: Disk | Partition | null,
+  expanded: Expanded | null,
+  actions: { children: string; onClick: () => void }[] = []
```

Comment:
Might be worth leaving a comment here about using the type from `TableMenu` when it gets migrated to TS.

Reply:
Good call!